DIPs/DIP1005.md

```diff
@@ -0,0 +1,781 @@
+
+Inline imports apply to all declarations. This includes the `with` declaration itself, having the consequence that multiple `with` import declarations may be applied in a cascading manner:
+
+```d
+with (import module_a : A)
+with (import module_b : B)
+A fun(B) { ... }
+```
+
+### Lookup rules
+
+When the name of a Dependency-Carrying Declaration is found via lookup, its corresponding inline imports are executed. Then the name is resolved.
+
+The visibility of the imported symbol(s) lasts through the end of the `with` declaration. That includes function contracts and bodies and also module constructors and inner functions.
+
+The inline imports have priority over existing `import`s visible to the declaration. This is so as to avoid other names present in the scope to have equal footing with names immediately present in the declaration. The lookup is equivalent to placing the inline `import`s in a scope unique to the declaration, where they take precedence in name resolution just like scoped imports per the current language rules. Example:
+
+```d
+import module_b;
+with (import module_a) void fun(X value) { ... }
+```
+
+The name `X` is looked up as if the code was structured as follows:
+
+```d
+import module_b;
+{
+    import module_a;
+    void fun(X value) { ... }
+}
+```
+
+This equivalent code, however, is not legal at top level. In that case we can artificially introduce an imaginary template to analyze lookup on compilable code:
+
+```d
+import module_b;
+template __unused()
+{
+    import module_a;
+    void fun(X value) { ... }
+}
+```
+
+The symbol `X` is looked up per the current language rules in the working code above.
+
+If two or more imports within the same `with` declaration define the same name, name resolution across these is the same as if the imports were top-level.
+
+If a module defines a symbol at top level and then imports a module, lookup proceeds similarly with local imports. Consider:
+
+```d
+int writeln;
+with (import std.stdio)
+void main() { writeln("hello, world"); }
+```
+This code is in error because `writeln` has type int. However, the following code is correct because it specifies the symbol name explicitly:
+
+```d
+int writeln;
+with (import std.stdio : writeln)
+void main() { writeln("hello, world"); }
+```
+
+### Examples
+
+Below are a few examples taken from the standard library:
+
+```d
+with (import std.meta, std.range, std.traits)
+auto uninitializedArray(T, I...)(I sizes) nothrow @system
+if (isDynamicArray!T && allSatisfy!(isIntegral, I) &&
+    hasIndirections!(ElementEncodingType!T))
+{
+    ...
+}
+```
+
+Alternatively, the declaration may specify the exact symbols needed by using multiple imports:
+
+```d
+with (import std.meta : allSatisfy)
+with (import std.range : ElementEncodingType)
+with (import std.traits : hasIndirections, isDynamicArray, isIntegral)
```

Comment:
This example would benefit a lot from [Issue 13855 – Allow multiple selective imports from different modules in a single import statement](https://issues.dlang.org/show_bug.cgi?id=13855) ;).